### ISSUE TITLE
Customize timezone

### ISSUE THREAD

**Issue by Frankwas353** [bromite/bromite#331]
**Is your feature request related to privacy?**
Yes - location (country) should be private.

**Is there a patch available for this feature somewhere?**
No.

**Describe the solution you'd like**
Javascript timezone "name" should return null.
The timezone offset can remain - it leaks only a large region ("europe") and not a country.

**Describe alternatives you've considered**
Changing system timezone in Android Settings.
This alternative breaks every android app that depends on time (calendar, reminders, etc) and has terrible UX (the phone doesn't show the correct time).

Currently, the following methods leak location (country/capital city of country):

```
window.Intl.DateTimeFormat().resolvedOptions().timeZone
"Africa/Johannesburg"

Date().toLocaleString('en-us',{timeZoneName:'long'})
"12:34:56 PM South Africa Standard Time"
```

The implementation seems to be here:
https://chromium.googlesource.com/v8/v8.git/+/refs/heads/master/src/objects/js-date-time-format.cc
in the function:
`MaybeHandle<JSObject> JSDateTimeFormat::ResolvedOptions()`

This line would probably fix the leak:
`timezone_value = factory->undefined_value();`

This line is already in the source code, in case resolving didn't work.

**Comment by Frankwas353** [comment]
In large countries it leaks the sub-region of the country, for example for Russia it leaks 1 of 24 sub-regions: Europe/Kaliningrad, Europe/Moscow, Europe/Volgograd, Asia/Yekaterinburg, Asia/Novosibirsk, etc...

**Comment by csagan5** [comment]
I estimate the breakage due to changing the timezone only there as immense; feel free to try a patch to figure it out.

Also, this is not the Tor Browser and the goal of Bromite has always been to avoid extremely precise private data leak; this is not an extremely precise bit of information.

**Comment by csagan5** [comment]
This could also be implemented by making some changes to `base/android/java/src/org/chromium/base/TimezoneUtils.java`.

**Comment by quark2323** [comment]
> This could also be implemented by making some changes to `base/android/java/src/org/chromium/base/TimezoneUtils.java`.

If anyone absolutely needs a different timezone you could replace this:
String timezoneID = TimeZone.getDefault().getID();
with this:
String timezoneID = "whatever_continent/whatever_city";

Ideally the timezone would be randomized on each session, or have it match your public IP's timezone. With a setting for this in Privacy.
I'm not good at coding but will try to make something for randomization. I think there's a function to call all available timezones.

**Comment by csagan5** [comment]
@quark2323 please report back your findings if you try this approach.

**Comment by quark2323** [comment]
Managed to randomize the timezone. It does it on each tab. 
`diff --git a/chromium/Vanadium/src/base/android/java/src/org/chromium/base/TimezoneUtils.java b/TimezoneUtils.java
index cddd3d9..164631f 100644
--- a/chromium/Vanadium/src/base/android/java/src/org/chromium/base/TimezoneUtils.java
+++ b/TimezoneUtils.java
@@ -11,6 +11,7 @@ import org.chromium.base.annotations.JNINamespace;
 import org.chromium.base.annotations.MainDex;
 
 import java.util.TimeZone;
+import java.util.Random;
 
 @JNINamespace("base::android")
 @MainDex
@@ -29,8 +30,11 @@ class TimezoneUtils {
         // On Android N or earlier, getting the default timezone requires the disk
         // access when a device set up is skipped.
         StrictMode.ThreadPolicy oldPolicy = StrictMode.allowThreadDiskReads();
-        String timezoneID = TimeZone.getDefault().getID();
+        String[] TimeZoneArray = TimeZone.getAvailableIDs();
+        Random random = new Random();
         StrictMode.setThreadPolicy(oldPolicy);
+        int randomItem = random.nextInt (TimeZoneArray.length);
+        String timezoneID = TimeZoneArray[randomItem];
         return timezoneID;
     }
 }
`
I don't know how to create a switch for randomizing in Privacy menu.

**Comment by csagan5** [comment]
@uazo are you interested in this feature request as well? See patch in previous comment.

I am asking you because you seem nimble with adding these settings; I was thinking it could be offered with 3 options:
* System default
* UTC
* Random

I am proposing this because adding a way to select from all timezones might be difficult. What do you think?

**Comment by uazo** [comment]
with great pleasure, but if you agree I would prefer to fix bugs before adding anything else.

**Comment by uazo** [comment]
a new content-settings would be better? so it would be possible to change it per site, who knows.

initialization appears to take place here
https://source.chromium.org/chromium/chromium/src/+/master:content/app/content_main_runner_impl.cc;l=738
so it's more per process than per tab (chromium may run in single process mode if resources are low)

here instead you could understand who are the clients who need the information, so as to understand where that information is used
https://source.chromium.org/chromium/chromium/src/+/master:services/device/time_zone_monitor/time_zone_monitor_android.cc

for example here is a blink method related to overriding the timezone
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/timezone/timezone_controller.cc;l=128

```
// An instance of TimeZoneController manages renderer time zone. It listens to
// the host system time zone change notifications from the browser and when
// received, updates the default ICU time zone and notifies V8 and workers.
//
// Time zone override mode allows clients to temporarily override host system
// time zone with the one specified by the client. The client can change the
// time zone, however no other client will be allowed to set another override
// until the existing override is removed. When the override is removed, the
// current host system time zone is assumed.
```
maybe we can use that, it needs to be analyzed better.

**Comment by csagan5** [comment]
I see that changing `TimezoneUtils.java` would not be sufficient.

> a new content-settings would be better? so it would be possible to change it per site, who knows.

Yeah, if it is not too much work, that would probably be the best, so that you either change it globally or per site. The majority of users will just change it globally though, so if we go with that solution it is already satisfactory.

> maybe we can use that, it needs to be analyzed better.

Yes, that is a good lead. You could also see how other settings are implemented (for example the one about toggling Javascript).

**Comment by uazo** [comment]
I almost finished the new feature and it seems to work.
@csagan5  do you think it is appropriate to change locale too?

**Comment by csagan5** [comment]
@uazo interesting, did not think of that. The locale part has a much bigger impact, I suggest to do it separately.

Reference: https://source.chromium.org/chromium/chromium/src/+/master:base/android/java/src/org/chromium/base/LocaleUtils.java

**Comment by uazo** [comment]
> https://source.chromium.org/chromium/chromium/src/+/master:base/android/java/src/org/chromium/base/LocaleUtils.java

ehm no.
if we touch that, also ui locale changes.

it's here:
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/inspector/locale_controller.h;l=16
for the renderer process

> I suggest to do it separately.

are you sure? add another content-setting?
I would want to completely modify that part of the upstream code (but I don't), it's messy and full of scattered if's ...

**Comment by ouzowtf** [comment]
I don't know how that works, but would a changed timezone affect the automatic darkening of the browser/websites in the evening?
Sorry for interrupting the dev talk 😅

**Comment by uazo** [comment]
> but would a changed timezone affect the automatic darkening of the browser/websites in the evening?

I don't know, I don't think that management is in charge of the rendering process, maybe @csagan5  knows more.
but when I see https://github.com/bromite/bromite/issues/716 I'll be able to tell you

**Comment by uazo** [comment]
> darkening of the browser/**websites**  ??

websites? also websites go in darken mode?
@ouzowtf  if it is the site itself via javascript to go into that mode, then in that case, yes, site might think it's night ..

**Comment by uazo** [comment]
> @ouzowtf if it is the site itself via javascript to go into that mode, then in that case, yes, site might think it's night ..

no, I'm stupid.
I only change offset of the timezone, not the time. So for sites via javascript time is correct but in UTC form.

**Comment by csagan5** [comment]
> if we touch that, also ui locale changes.
> 
> it's here:
> https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/inspector/locale_controller.h;l=16
> for the renderer process
> 
> > I suggest to do it separately.
> 
> are you sure? add another content-setting?
> I would want to completely modify that part of the upstream code (but I don't), it's messy and full of scattered if's ...

How are you implementing it for the timezone? I do not know that.

**Comment by ouzowtf** [comment]
@uazo I didn't mean when the website itself has some JavaScript to provide that, but what Chrome/Chromium provides as an option. Maybe related to #716.
Nevermind, this question just striked my brain by reading this issue :joy:

**Comment by uazo** [comment]
> How are you implementing it for the timezone? I do not know that.

https://gist.github.com/uazo/8536dcb8483b157fcbc489acc13ceae5

**Comment by csagan5** [comment]
> https://gist.github.com/uazo/8536dcb8483b157fcbc489acc13ceae5

@uazo it looks very good, if you open a PR I can leave my feedback; so your idea would be to derive the locale from the timezone? I am against that: the user has to change the Android system language to change the locale. If only the locale is overridden you will still get a difference in the `Accept-Language` header which is sent vs `navigator.language` from Javascript

**Comment by uazo** [comment]
> @uazo it looks very good, if you open a PR I can leave my feedback; 

I have to finish it.
random (easy) is missing and I wanted to add the possibility to define a textbox in chrome://flags with the offset set by the user (I saw that it seems possible).

> the user has to change the Android system language to change the locale. If only the locale is overridden you will still get a difference in the Accept-Language header which is sent vs navigator.language from Javascript

maybe you can take it off in that case. I had already seen (with cookies fix) that requests are all managed in a single precise point.
 Accept-Language header: for the standard is mandatory?

**Comment by csagan5** [comment]
> I wanted to add the possibility to define a textbox in chrome://flags with the offset set by the user (I saw that it seems possible).

You mean a combo box with offsets for the timezones? But timezones are normally not defined like that, it would be better to offer a list of timezones, see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

The `GMT +02:00` version does not happen normally on any browser; it is valid but only used for development.



> maybe you can take it off in that case. 

I'll let you finish the patch but I am not in favor of changing the locale as well, it is a can of worms.

> I had already seen (with cookies fix) that requests are all managed in a single precise point.

That was good experience then!



> Accept-Language header: for the standard is mandatory?

Not sure what standard you are referring to, but all browser navigation requests have it and we cannot omit it anyways.

**Comment by uazo** [comment]
> You mean a combo box with offsets for the timezones? 

I don't know yet. Combo or TextBox.

> But timezones are normally not defined like that, it would be better to offer a list of timezones, see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

Ok, I follow that list

> The GMT +02:00 version does not happen normally on any browser; it is valid but only used for development.

yes, but can it be used to track bromite browser?

>  I am not in favor of changing the locale as well, it is a can of worms.

I don't know.

> Not sure what standard you are referring to

https://tools.ietf.org/html/rfc7231#section-5.3.5
it seems optional

**Comment by csagan5** [comment]
> Ok, I follow that list

This is a better list: https://source.chromium.org/chromium/chromium/src/+/master:third_party/perl/perl/vendor/lib/DateTime/TimeZone/

However I think it's not a good idea to embed this list, we need a free text or just pick the most popular timezones for a combo box.

> yes, but can it be used to track bromite browser?

Yes.

**Comment by uazo** [comment]
> > @ouzowtf if it is the site itself via javascript to go into that mode, then in that case, yes, site might think it's night ..

> no, I'm stupid.
I only change offset of the timezone, not the time. So for sites via javascript time is correct but in UTC form.

@csagan5  I was (re)thinking about this.
It is true that for the whole world the UTC time is always the same, but the sites, not knowing the offset, do not know if it is day or night in the user's location.

**Comment by csagan5** [comment]
> @csagan5 I was (re)thinking about this.
> It is true that for the whole world the UTC time is always the same, but the websites, not knowing the offset, do not know if it is day or night at the user's location.

Timezone does not change depending on the hour of the day. Please let's not discuss dark themes or website darkening here on this issue.

**Comment by uazo** [comment]
>  Please let's not discuss dark themes or website darkening here on this issue.

sure, but I'm talking in general about all the website javascript features, based on timezone offset (like that) affected by the override change.
we have to warn the user that they will be broken.

**Comment by csagan5** [comment]
> sure, but I'm talking in general about all the website javascript features, based on timezone offset (like that) affected by the override change.
> we have to warn the user that they will be broken.

They will not be broken, they will return time based on the timezone that the user has selected. It is sufficient to explain what changing the timezone does in the description of the setting/flag.

**Comment by csagan5** [comment]
Fixed in `85.0.4183.94`.
